sql: CockroachDB crashes when executing the PoC query.

**Describe the problem**

The latest version of CockroachDB (d39558f51d) crashes when executing the following query: 

```sql
SET testing_optimizer_disable_rule_probability = 1.0;
WITH ta11 (ca12) AS (SELECT '{00000000000000000000000000000000000000000000000000000000"cjhVWKAbt":9}' || n FROM ROWS FROM (generate_series(7287339425617371612, 9223372036854775807)) AS ta10 (n)) SELECT * FROM ta11 WHERE ca12 = '11';
```

**To Reproduce**

Here is the detail steps to reproduce the bug. 

1. In system Ubuntu 20.04, download the CockroachDB source code from github. 
2. Use the latest version of the CockroachDB code (tested version: d39558f51d). 
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment. 
5. Observe the crash. 

```sql
# PoC
SET testing_optimizer_disable_rule_probability = 1.0;
WITH ta11 (ca12) AS (SELECT '{00000000000000000000000000000000000000000000000000000000"cjhVWKAbt":9}' || n FROM ROWS FROM (generate_series(7287339425617371612, 9223372036854775807)) AS ta10 (n)) SELECT * FROM ta11 WHERE ca12 = '11';
```

Stack Trace: 
```
warning: error retrieving the database name: ERROR: use of crdb_internal_vtable_pk column not allowed. (SQLSTATE XXUUU)
*
* ERROR: a SQL panic has occurred while executing the following statement:
* WITH ta11 (ca12) AS (SELECT '{00000000000000000000000000000000000000000000000000000000"cjhVWKAbt":9}' || n FROM ROWS FROM (generate_series(7287339425617371612, 9223372036854775807)) AS ta10 (n)) SELECT * FROM ta11 WHERE ca12 = '11'
*
*
* ERROR: a panic has occurred!
* buffer 1 (ta11)-limited: no bytes to release, current 0, free 196336
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:826
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:804
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:644
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:556
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:132
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1303
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1582
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1222
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:731
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2423
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:128
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1922
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1927
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1847
* Wraps: (2) while executing: WITH _ (_) AS (SELECT '_' || _ FROM ROWS FROM (generate_series(_, _)) AS _ (_)) SELECT * FROM _ WHERE _ = '_'
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:826
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:804
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:644
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:556
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:132
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1303
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1582
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1222
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:731
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2423
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:128
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1922
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1927
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1847
* Wraps: (4) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:804
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:644
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:556
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:132
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1303
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1582
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1222
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:731
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2423
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:128
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1922
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1927
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1847
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:828
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728
*   | runtime.goexit
*   | 	/usr/local/go/src/runtime/asm_amd64.s:1594
* Wraps: (5) buffer 1 (ta11)-limited: no bytes to release, current 0, free 196336
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *withstack.withStack (5) *errutil.leafError
*
panic: buffer 1 (ta11)-disk: unexpected 52428800 leftover bytes
	panic: buffer 1 (ta11)-limited: no bytes to release, current 0, free 196336 [recovered]
	panic: buffer 1 (ta11)-limited: no bytes to release, current 0, free 196336

goroutine 4670 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc004cbf900, {0x804ea50, 0xc0054a5c00}, {0x4ed64c0?, 0xc005854150?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1071 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x804ea50, 0xc0054a5c00}, {0x2f682d12?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:826 +0x4a
panic({0x4ed64c0, 0xc005854150})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x804eaf8, 0xc005d82ff0}, 0xc0028bc000, {0x53ad849?, 0x1?}, {0xc005873d60?, 0xc00093d788?, 0xc005873d30?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378 +0x1c5
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes(0xc0023059a0, {0x804eaf8, 0xc005d82ff0}, 0x2fef0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:804 +0x1a5
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close(0xc001ab99c0?, {0x804eaf8?, 0xc005d82ff0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:644 +0x45
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close(0xc002908c00, {0x804eaf8, 0xc005d82ff0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:556 +0xc5
github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close(0xc00239db28, {0x804eaf8, 0xc005d82ff0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:132 +0x46
github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close(0xc00239db00, {0x804eaf8, 0xc005d82ff0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67 +0x4b
github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close(0xc005723790, {0x804eaf8, 0xc005d82ff0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382 +0x3c
github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close(0xc004cc0528, {0x804eaf8, 0xc005d82ff0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451 +0x74
github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close(0x800e680?, {0x804eaf8?, 0xc005d82ff0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475 +0x67
panic({0x4ed64c0, 0xc005854108})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x804eaf8, 0xc005d83950}, 0xc0028bc000, {0x534887b?, 0x33881fa?}, {0xc005874198?, 0x1?, 0x1?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:378 +0x1c5
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc002305a40, {0x804eaf8, 0xc005d83950}, 0x1)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445 +0x21a
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close(0xc00239db28, {0x804eaf8, 0xc005d83950})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134 +0x85
github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Close(0xc00239db00, {0x804eaf8, 0xc005d83950})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/buffer.go:67 +0x4b
github.com/cockroachdb/cockroach/pkg/sql.(*planMaybePhysical).Close(0xc005723790, {0x804eaf8, 0xc005d83950})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:382 +0x3c
github.com/cockroachdb/cockroach/pkg/sql.(*planComponents).close(0xc004cc0528, {0x804eaf8, 0xc005d83950})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:451 +0x74
github.com/cockroachdb/cockroach/pkg/sql.(*planTop).close(0x1?, {0x804eaf8?, 0xc005d83950?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:475 +0x67
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc004cbf900?, {0x804eaf8, 0xc005d83950}, 0xc0051f0800?, 0xc004cbfdd0?, 0xc004cbfdd0, 0xc002eed880, 0x3?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1303 +0x3b4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004cbf900, {0x804eaf8, 0xc005d83950}, 0xc004cbfdd0, 0xc005d83950?, {0x7f229418daf8?, 0xc002bacad8}, 0x7c?, 0xc000dff8d8)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1582 +0x4c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004cbf900, {0x804eaf8, 0xc005d82ff0}, 0xc004cbfdd0, {0x7f229418daf8, 0xc002bacad8})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1222 +0xbcf
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004cbf900, {0x804eaf8?, 0xc005d82ed0?}, {{0x8077800, 0xc0057234a0}, {0xc002fe2350, 0xe7}, 0x0, 0x1}, 0x0, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:731 +0x24b4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x804eaf8?, 0xc005d82ed0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0xfc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x804eaf8?, {0x804eaf8?, 0xc005d82ed0?}, {0x8077800?, 0xc0057234a0?}, 0x0?, 0x0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2423 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004cbf900, {0x804eaf8, 0xc005d82ed0}, {{0x8077800, 0xc0057234a0}, {0xc002fe2350, 0xe7}, 0x0, 0x1}, 0x0, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:128 +0x4bf
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1({{{0x8077800, 0xc0057234a0}, {0xc002fe2350, 0xe7}, 0x0, 0x1}, {0xc0ceee4100f4c2b3, 0xd2a4067c, 0x0}, {0xc0ceee4100f4c500, ...}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1922 +0x2ff
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004cbf900)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1927 +0x1128
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004cbf900, {0x804ea50, 0xc0054a5c00}, 0x4187cc8a70000000?, 0xc004ac1dd8?, 0xc00529f6a0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1847 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc001976aa0?, {0x804ea50?, 0xc0054a5c00?}, {0xc0022ffc00?}, 0x3?, 0xc0039a78a0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:828 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728 +0x40d
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:639 +0x22e
```


**Expected behavior**
The CockroachDB server should continue working, even if the query contains semantic errors. 

**Additional data / screenshots**
The PoC query: (minimized) 

```sql
SET testing_optimizer_disable_rule_probability = 1.0;
WITH ta11 (ca12) AS (SELECT '{00000000000000000000000000000000000000000000000000000000"cjhVWKAbt":9}' || n FROM ROWS FROM (generate_series(7287339425617371612, 9223372036854775807)) AS ta10 (n)) SELECT * FROM ta11 WHERE ca12 = '11';
```

**Environment:**
 - CockroachDB version: v22.2.0-alpha.3-1592-gd39558f51d
 - Server OS: Ubuntu Server 20.04 LTS
 - Client app: Cockroach demo command. 

**Additional context**
Cockroach crashes. 


Jira issue: CRDB-20955

--- a/pkg/sql/rowcontainer/hash_row_container.go
+++ b/pkg/sql/rowcontainer/hash_row_container.go
@@ -886,6 +886,15 @@ func (h *HashDiskBackedRowContainer) SpillToDisk(ctx context.Context) error {
 		return errors.New("already using disk")
 	}
 	hdrc := MakeHashDiskRowContainer(h.diskMonitor, h.engine)
+	defer func() {
+		if h.src != &hdrc {
+			// For whatever reason, we weren't able to spill, so in order to not
+			// leak any of the disk resources we explicitly close the disk
+			// container we've just created - otherwise, we will lose the
+			// reference to it and won't ever close it properly.
+			hdrc.Close(ctx)
+		}
+	}()
 	if err := hdrc.Init(ctx, h.shouldMark, h.types, h.storedEqCols, h.encodeNull); err != nil {
 		return err
 	}

--- a/pkg/sql/rowcontainer/row_container.go
+++ b/pkg/sql/rowcontainer/row_container.go
@@ -592,6 +592,8 @@ func (f *DiskBackedRowContainer) SpillToDisk(ctx context.Context) error {
 		return errors.New("already using disk")
 	}
 	drc := MakeDiskRowContainer(f.diskMonitor, f.mrc.types, f.mrc.ordering, f.engine)
+	f.src = &drc
+	f.drc = &drc
 	if f.deDuplicate {
 		drc.DoDeDuplicate()
 		// After spilling to disk we don't need this map to de-duplicate. The
@@ -616,9 +618,6 @@ func (f *DiskBackedRowContainer) SpillToDisk(ctx context.Context) error {
 		}
 	}
 	f.mrc.Clear(ctx)
-
-	f.src = &drc
-	f.drc = &drc
 	f.spilled = true
 	return nil
 }

